Objects/funcobject.c

```diff
@@ -678,18 +678,23 @@ static int
 func_clear(PyFunctionObject *op)
 {
     op->func_version = 0;
-    Py_CLEAR(op->func_code);
     Py_CLEAR(op->func_globals);
     Py_CLEAR(op->func_builtins);
-    Py_CLEAR(op->func_name);
-    Py_CLEAR(op->func_qualname);
     Py_CLEAR(op->func_module);
     Py_CLEAR(op->func_defaults);
     Py_CLEAR(op->func_kwdefaults);
     Py_CLEAR(op->func_doc);
     Py_CLEAR(op->func_dict);
     Py_CLEAR(op->func_closure);
     Py_CLEAR(op->func_annotations);
+    /* Below are required attributes, so to keep everything in
+     * a consistent state, don't clear them.  They're immutable,
+     * so they can't participate in cycles anyway.
```

Comment:
Did you confirm that Unicode subclass are not allowed?

Reply:
Ah, good catch, it can be a subclass:

```python
>>> def f():
...     pass
...
>>> class S(str):
...     pass
...
>>> f.__name__ = S("???") # no error
```

I don't know what a good solution is.